go/arrow/bitutil/bitmaps.go

```diff
@@ -422,3 +423,81 @@ func CopyBitmap(src []byte, srcOffset, length int, dst []byte, dstOffset int) {
 	dst[nbytes-1] &= ^trailMask
 	dst[nbytes-1] |= lastData & trailMask
 }
+
+type bitOp struct {
+	opWord func(uint64, uint64) uint64
+	opByte func(byte, byte) byte
+}
+
+var (
+	bitAndOp = bitOp{
+		opWord: func(l, r uint64) uint64 { return l & r },
+		opByte: func(l, r byte) byte { return l & r },
+	}
+	bitOrOp = bitOp{
+		opWord: func(l, r uint64) uint64 { return l | r },
+		opByte: func(l, r byte) byte { return l | r },
+	}
+)
+
+func alignedBitmapOp(op bitOp, left, right []byte, lOffset, rOffset int64, out []byte, outOffset int64, length int64) {
+	debug.Assert(lOffset%8 == rOffset%8, "aligned bitmap op called with unaligned offsets")
+	debug.Assert(lOffset%8 == outOffset%8, "aligned bitmap op called with unaligned output offset")
+
+	nbytes := BytesForBits(length + lOffset%8)
+	left = left[lOffset/8:]
+	right = right[rOffset/8:]
+	out = out[outOffset/8:]
+	for i := int64(0); i < nbytes; i++ {
+		out[i] = op.opByte(left[i], right[i])
+	}
+}
+
+func unalignedBitmapOp(op bitOp, left, right []byte, lOffset, rOffset int64, out []byte, outOffset int64, length int64) {
+	leftRdr := NewBitmapWordReader(left, int(lOffset), int(length))
+	rightRdr := NewBitmapWordReader(right, int(rOffset), int(length))
+	writer := NewBitmapWordWriter(out, int(outOffset), int(length))
+
+	for nwords := leftRdr.Words(); nwords > 0; nwords-- {
+		writer.PutNextWord(op.opWord(leftRdr.NextWord(), rightRdr.NextWord()))
```

Comment:
Interesting, I only copied the C++ implementation versions of `alignedBitmapOp` and `unalignedBitmapOp`. I'll add a benchmark for this and see what I discover

Reply:
You're completely correct, `unalignedBitmapOp` is faster! i'll make some modifications to get the aligned one to work word-wise where it can. This might be a good spot to also improve the C++ implementation which would show the same thing (as it also works byte-wise for the aligned case)

Reply:
Before:

```
goos: linux
goarch: amd64
pkg: github.com/apache/arrow/go/v10/arrow/bitutil
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkBitmapAnd/nbytes=32768/aligned-20 	   25418	     48190 ns/op	1359.95 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=32768/unaligned-20         	   29241	     40694 ns/op	1610.46 MB/s	     192 B/op	       3 allocs/op
BenchmarkBitmapAnd/nbytes=131072/aligned-20          	    6798	    194915 ns/op	1344.91 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=131072/unaligned-20        	    7018	    162163 ns/op	1616.55 MB/s	     192 B/op	       3 allocs/op
```

After:

```
goos: linux
goarch: amd64
pkg: github.com/apache/arrow/go/v10/arrow/bitutil
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkBitmapAnd/nbytes=32768/aligned-20 	  126613	      9308 ns/op	7040.73 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=32768/unaligned-20         	   28724	     41392 ns/op	1583.30 MB/s	     192 B/op	       3 allocs/op
BenchmarkBitmapAnd/nbytes=131072/aligned-20          	   32292	     37420 ns/op	7005.50 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=131072/unaligned-20        	    7068	    166973 ns/op	1569.97 MB/s	     192 B/op	       3 allocs/op
```

Reply:
Well, if I run the C++ benchmarks, I get this:
```
BenchmarkBitmapAnd/32768/0                   1696 ns         1696 ns       411775 bytes_per_second=17.9927G/s
BenchmarkBitmapAnd/131072/0                  6843 ns         6842 ns       101856 bytes_per_second=17.8405G/s
BenchmarkBitmapAnd/32768/1                   3999 ns         3998 ns       174766 bytes_per_second=7.63287G/s
BenchmarkBitmapAnd/131072/1                 15546 ns        15544 ns        44802 bytes_per_second=7.85325G/s
BenchmarkBitmapAnd/32768/2                   3988 ns         3987 ns       175050 bytes_per_second=7.65504G/s
BenchmarkBitmapAnd/131072/2                 15601 ns        15599 ns        44678 bytes_per_second=7.82568G/s
```
... where the first benchmark parameter is the number of bytes and the second is the bif offset, so you see the results show a similar performance profile as the Go ones (with the aligned implementation being a couple times faster than the unaligned one).

The reason this works is probably that C++ compilers optimize hard enough to turn the byte-wise loop into a word-wise loop plus byte-wise prolog/epilog.

Reply:
That makes a lot of sense, I could probably use the same trick I've used before by writing the loop in C, compiling to asm and converting to golang ASM like we do for SIMD. I'll take a stab at that since that would simplify the logic here significantly.

Reply:
Well this is much better...

```
goos: linux
goarch: amd64
pkg: github.com/apache/arrow/go/v10/arrow/bitutil
cpu: 12th Gen Intel(R) Core(TM) i7-12700H
BenchmarkBitmapAnd/nbytes=32768/0-20       	 1110499	      1078 ns/op	60797.80 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=32768/1-20       	   29881	     39699 ns/op	1650.82 MB/s	     192 B/op	       3 allocs/op
BenchmarkBitmapAnd/nbytes=32768/2-20       	   30028	     39396 ns/op	1663.50 MB/s	     192 B/op	       3 allocs/op
BenchmarkBitmapAnd/nbytes=131072/0-20      	  254810	      4698 ns/op	55795.91 MB/s	       0 B/op	       0 allocs/op
BenchmarkBitmapAnd/nbytes=131072/1-20      	    7418	    159083 ns/op	1647.85 MB/s	     192 B/op	       3 allocs/op
BenchmarkBitmapAnd/nbytes=131072/2-20      	    7378	    157388 ns/op	1665.59 MB/s	     192 B/op	       3 allocs/op
```

:smile: Though it looks like i should make a card to look into similarly speeding up the unaligned cases (by way of making the bitmapword reader and writer faster)

Reply:
Well, I took a look at the C++ benchmark, and it's actually limited by the popcount that's done at the end :-)
If I remove the popcount I get this:
```
BenchmarkBitmapAnd/32768/0         732 ns          732 ns       967465 bytes_per_second=41.6736G/s
BenchmarkBitmapAnd/131072/0       3105 ns         3105 ns       229726 bytes_per_second=39.3198G/s
BenchmarkBitmapAnd/32768/1        2913 ns         2913 ns       240233 bytes_per_second=10.4774G/s
BenchmarkBitmapAnd/131072/1      11528 ns        11526 ns        60865 bytes_per_second=10.5912G/s
BenchmarkBitmapAnd/32768/2        2924 ns         2924 ns       236873 bytes_per_second=10.4378G/s
BenchmarkBitmapAnd/131072/2      11552 ns        11550 ns        60619 bytes_per_second=10.5691G/s
```